### ISSUE TITLE
Remerge Dave's branch?

### ISSUE THREAD

**Issue by oberstet** [crossbario/autobahn-cpp#50]
Consider remerging https://github.com/davidchappelle/AutobahnCpp

**Comment by davidchappelle** [comment]
Just to throw this out there as well. If you do not feel that my fork has gone in the direction that you would like to have seen would you be open to allowing for a project renaming and copyright change?

**Comment by oberstet** [comment]
ah, shit. yes, I still have to respond to your last mail (I now remember). I will do _now_.

rgd copyright change: what do you mean? switch license?

**Comment by ecorm** [comment]
> would you be open to allowing for a project renaming and **copyright change**?

I hope you meant **adding** yourself (or your company) as a copyright holder. As in:

```
//  Copyright (C) 2014 Tavendo GmbH
//  Copyright (C) 2015 Other Contributor
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  ...
```

Or, as suggested in [this Q&A](http://programmers.stackexchange.com/questions/157968):

```
//  Original work Copyright (C) 2014 Tavendo GmbH
//  Modified work Copyright (C) 2015 Other Contributor
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  ...
```

**Comment by oberstet** [comment]
@davidchappelle 

Yes, Emil is right indeed. Removing the "Tavendo GmbH" line isn't something we would support, and to be honest, I don't see a reason why you would want to do that?

I would support changing the license from Apache 2.0 to MIT - but no other license. For this, we would need to ask **all** contributors for permission. The reason for going MIT is: Apache 2.0 is incompatible with GPL 2. We've gone through that change for AutobahnJS and AutobahnPython already.

Note that neither we (Tavendo) nor anybody else can relicense or dual-license AutobahnCpp or derivative works (as any AutobahnCpp fork): for this we would have needed a contributor license agreement signed by all contributors *before* merging any code - and we (deliberately) haven't done that.

**Comment by oberstet** [comment]
Now to the technical matter .. following up on our discussion here https://groups.google.com/forum/#!topic/wampws/GryF1YHc_go

I think I can get over the splitting up into many small files;) I don't like it a lot, but so what. Works.

The main concern for me is really the public API. Rereading the thread and the actual code in the fork, I don't get it:

You say, boost::any does malloc even on an integer. that's generally to be avoided (any malloc) if possible and sane. I follow that completely.

You say msgpack::object is a variant type just by accident also provided by the msgpack library, and does not "leak" msgpack as the only serialization possible into the public API. I can follow that argument - and would be totally convinced if there would be a second supported serialization format. If so, the we would just say: msgpack::object is better than boost::any, but under the hood, msgpack::object could be filled from JSON as well (which would then probably use a JSON library just serializing to msgpack::object

What I don't get is: https://github.com/davidchappelle/AutobahnCpp/blob/master/examples/caller.cpp#L68

This is a plain old STL container. I thought the public API is about msgpack::object? Or is that magically coerced from STL tuple to msgpack::object?

**Comment by oberstet** [comment]
@ecorm : on a higher level, I am slightly concerned about the "proliferation" of user options with for C++ _clients_.

I mean, recently s.o. announced work on WAMP/C++ for Qt. This is fine, as it strives to deeply integrate into the Qt event loop.

Boost/ASIO is another event loop vastly different - so it probably deserves it's own WAMP library.

But CppWAMP also is based on ASIO (only)?

IOW: should we really move forward with like 3 flavors of WAMP client library based on ASIO? Just thinking loud here ..

**Comment by ecorm** [comment]
@oberstet Remember that I had originally considered contributing to Autobahn|Cpp, as per this group [discussion](https://groups.google.com/d/topic/autobahnws/WzS4TZIDkZQ/discussion). I had concluded that we had different design philosophies, and "negotiating" changes to the Autobahn|Cpp API would take too much time.

I didn't realize it the time, but the experimental nature of `boost::future` ended up being a deal breaker for me. I now prefer the callback and co-routine based API of CppWAMP. And CppWAMP is architected in a way that a future-based API can be easily layered on top of the async API.

I would have welcomed contributions from Dave, but he decided instead to fork Autobahn|Cpp.

I say let people freely innovate and create their own C++ implementations if they choose. Ultimately, users will vote with their feet.

**Comment by oberstet** [comment]
> Remember that I had originally considered contributing ..

Ah yeah, right;) Sorry. All of this is moving at a pace which makes it sometimes hard to follow / remember everything.

>  I now prefer the callback and co-routine based API ..

Fair enough. Note to self: I really have to look how co-routines are done in C++ ..

> I say let people freely innovate and create their own C++ implementations if they choose. Ultimately, users will vote with their feet.

Yes, I agree. And you just remembered me of the very reason why at least 2 C++/ASIO/WAMP libs: std::future vs callbacks vs co-routines ...

My personal view: callbacks is a no go for me. I prefer std::future. BUT: co-routines, that is interesting ..

**Comment by oberstet** [comment]
fwiw, seems like boost::coroutine is also a std proposal http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3985.pdf - at least same author ..

**Comment by davidchappelle** [comment]
*Libraries*

It would be nice to only have 1 C++ WAMP library based on asio. For my
purposes, coroutines were not something the engineers I am working with
were going to get behind. So unfortunately I had to rule out CppWAMP. In
the end I decided that it was best to move forward with AutobahnCpp and try
to move that project forward. Although, boost::future is still experimental
it is where STL will eventually pull its implementation from so that is a
plus.

*Licensing*

I was just suggesting that if my fork was not going to get merged back to
AutobahnCpp, would it be possible to rename the project so as to detach it
from AutobahnCpp. The copyright as it is now is fine.

*Technical Details*

> I think I can get over the splitting up into many small files;) I don't
like it a lot, but so what. Works.

I like to follow the 1 class per header file rule. It helps a fair bit with
maintainability.

> The main concern for me is really the public API. Rereading the thread
and the actual code in the fork, I don't get it:

> You say, boost::any does malloc even on an integer. that's generally to
be avoided (any malloc) if possible and sane. I follow that completely.

Yes. Msgpack is optimized for these intrinsic use cases. Furthermore, it
does not go to the system heap for all allocations which makes its
allocations much faster.

> You say msgpack::object is a variant type just by accident also provided
by the msgpack library, and does not "leak" msgpack as the only
serialization possible into the public API. I can follow that argument -
and would be totally convinced if there would be a second supported
serialization format. If so, the we would just say: msgpack::object is
better than boost::any, but under the hood, msgpack::object could be filled
from JSON as well (which would then probably use a JSON library just
serializing to msgpack::object

Msgpack actually advertises msgpack::object as an intentional variant type.
The intention is to eventually add json serialization support. Likely in
the same manner that was done for bonefish.

> What I don't get is:
https://github.com/davidchappelle/AutobahnCpp/blob/master/examples/caller.cpp#L68

> This is a plain old STL container. I thought the public API is about
msgpack::object? Or is that magically coerced from STL tuple to
msgpack::object?

Msgpack provides object serialization support for all of the STL container
types. Tuples, vectors, arrays, etc are represented as lists. Maps,
hashmaps, etc are represented as maps. These translate nicely to the wire
format specifications for WAMP messages. The API is templated such that it
can accept things like tuples, vectors, arrays, maps, etc. Which I find
allows the user to choose a container type that is appropriate to their
chosen data representation. Perhaps some comments and 'enable_if'
statements are needed.

**Comment by davidchappelle** [comment]
Curious what the user space context switching overhead of coroutines is.

**Comment by davidchappelle** [comment]
FWIW @ecorm basically implemented his own variant type for CppWAMP.
However, I think msgpack::object is still more flexible and more efficient.
It also has the added benefit of already supporting a huge number of types
by default. Pretty much the entire STL container library. It also has very
active contributions from a wide range of platforms/compilers.

**Comment by oberstet** [comment]
@davidchappelle 

Ok, let's make a decision. I am ready to merge back your fork and follow and actively support the decisions made: using msgpack::object in the public API, using CMake instead of SCons, source code organization.

There is only 1 thing left: I'd like to make the change to MIT. This requires your support and support by everyone that contributed to your fork.

I think having both AutobahnCpp and CppWamp continue is fine, as they made different decisions / tradeoffs .. and there isn't one that's "right" and the other wrong. Let users choose.

Deal?

**Comment by oberstet** [comment]
Oh, rgd license: just noticed msgpack is Apache (https://github.com/msgpack/msgpack-c/blob/master/LICENSE) .. 

So AutobahnCpp couldn't be used in GPL2 code even when it's MIT.

License crap.

Is Boost license GPL2 compatible btw? If so, I'd be fine with that also. The main thing is: allow use in GPL2 code ..

**Comment by davidchappelle** [comment]
Looks like Boost is listed as compatible with GPL v2:

http://gplv3.fsf.org/wiki/index.php/Compatible_licenses

And Apache v2 is listed as compatible with GPL v3

**Comment by jpetso** [comment]
My personal impression about AutobahnCpp, when I first saw it, was that it's a fairly minimal client that only adds extra code and dependencies where really necessary. That makes it both maintainable and easy to get into. I'm undecided whether I prefer @oberstet's all-in-one-file approach or @davidchappelle's one-file-per-class, but even after that split it's straightforward to read and contribute to. CppWAMP in that respect is more modular, and as a result offers more options but it's also more complex, with more concepts to understand. Maybe Dave's branch will develop into that same direction over time, I don't know.

In that sense, I like the dependency on futures. They've been in C++ since '11 and seem like more of a "core" feature that will underlie a number of other features. One could probably implement callbacks on top of futures and futures on top of callbacks, in the end those are pretty interchangeable for all practical purposes. The main difference being that callbacks require some sort of event loop integration and futures using mutexes/condvars underneath, plus exceptions are really nicely handled by futures imho.

Either way, I think there's room for a library that focuses on just one option and fully embraces it, with everything else tacked on top. For AutobahnCpp, futures and msgpack would be the underlying infrastructure and JSON can be that optional piece. For CppWAMP, callbacks and a custom variant type are the underlying infrastructure and JSON/msgpack/coroutines/futures are extensions.

@oberstet, are you looking at the license change just to make it work with GPLv2? Are you actually looking to intermingle it with GPLv2 or only use it from there? Because, as you yourself described beautifully in issue #16, Apache won't get viral and eat into GPL code because it's a separate library that only copylefts code in the AutobahnCpp code itself but no application code or anything.

**Comment by oberstet** [comment]
@davidchappelle

Ok, so I'd be fine with changing AutobahnCpp to either MIT or Boost.

But then it still coudn't be used in GPL2 software .. the FSF explicitly mentions the Apache 2.0 as being incompatible with the GPL2 (though it's compatible with the GPL3).

Oh well. We probably just can't fix that.

@jpetso

> are you looking at the license change just to make it work with GPLv2

That is my only itch.

> A minor one would be it's nice to say: _all_ Autobahn's are MIT. But that's cosmetics.

**Comment by oberstet** [comment]
Screw it: I am fine if we stick to Apache 2.0. We can't fix the GPL2.

@davidchappelle so I guess I just merge?

**Comment by davidchappelle** [comment]
That sounds great Tobias!

**Comment by oberstet** [comment]
Could you do me a favor? Create a new PR that I can just cleanly merge .. I am a bit "under water" with so many things in parallel ..

**Comment by jpetso** [comment]
[http://www.apache.org/licenses/GPL-compatibility.html](Apache says):

> This licensing incompatibility applies only when some Apache project software becomes a derivative work of some GPLv3 software, because then the Apache software would have to be distributed under GPLv3. This would be incompatible with ASF's requirement that all Apache software must be distributed under the Apache License 2.0.
>
> We avoid GPLv3 software because merely linking to it is considered by the GPLv3 authors to create a derivative work. We want to honor their license. Unless GPLv3 licensors relax this interpretation of their own license regarding linking, our licensing philosophies are fundamentally incompatible. This is an identical issue for both GPLv2 and GPLv3.

I don't believe there's yet a court ruling that conclusively sorts out what the GPL can consider a "derivative work" and what it can't, but it seems fair to avoid legal gray zones. If you want to ask the msgpack authors to consider a switch to the Boost license, I'll support that request. Not sure how diligently they would handle such a request though, especially since they've had a few more contributors than AutobahnCpp and only recently started to add copyright lines other than the original author's one. From the Topology side, I'm sure we wouldn't have a problem with switching to Boost - if you would like to go that direction, let's not put it off for too long.

**Comment by ecorm** [comment]
FWIW, CppWAMP *does* have experimental support for boost::future: https://github.com/ecorm/cppwamp/issues/3

I'm waiting for recent developments in Boost.Asio and Boost.Thread to be released, as described in the issue I referenced.

**Comment by oberstet** [comment]
@jpetso I agree, unless and until there is precedence from court ruling, it's basically unclear. But at least the FSF clearly states that GPL2 and Apache 2.0 are incompatible (http://gplv3.fsf.org/wiki/index.php/Compatible_licenses#GPLv2-incompatible_licenses)

We have been approached by other projects rgd AutobahnJS and AutobahnPython: e.g. Buildbot (a GPL2 CI system) will be using AutobahnPython upcoming .. and we've gone through a license change so they can use it.

So I think it would be nice if AutobahnCpp had a GPL2 compatible license _and_ msgpack too. If only one of these makes a switch, it doesn't make a difference for users: they still have a problem if they are on GPL2.

If you think you can ask and make msgpack change license, then let's do it too! And yeah, do it soon.

I am fine with MIT, and I am fine with Boost. The latter is probably even nicer, since then a user that "only" uses Boost, AutobahnCpp and msgpack just has to look at 1 license type.

If msgpack doesn't want to change license, then let's stick with Apache for AutobahnCpp. Because it doesn't make a difference anyway ..

**Comment by oberstet** [comment]
As a matter of fact, I just merged! Wooaa. ;) Thanks Topology, thanks Dave!

We can sort out the license "issue" still .. I merged since I am fine with and support any of above mentioned license options.

**Comment by jpetso** [comment]
Yay! Note that the filename and contents of the examples changed, so we had to adapt the links to not point to the tavendo repository. Those should now be switched back.

**Comment by davidchappelle** [comment]
I will try and go through the README.md as well and take a pass at updating
the coding examples.

**Comment by oberstet** [comment]
Awesome! Thx.

Am 14.09.2015 9:13 nachm. schrieb davidchappelle <notifications@github.com>:
I will try and go through the README.md as well and take a pass at updating
the coding examples.

**Comment by davidchappelle** [comment]
I guess this can be closed off for now.
